### DISCUSSION TITLE
Which extensions are important for engine operation?

### DISCUSSION THREAD

**Discussion by 10100010010001001010** [axmolengine/axmol#1081] [General]
Which extensions are important for engine operation? After deleted all of them i have linkage error with some libs from build\lib\debug

**Comment by rh101** [comment]
You can't just delete files randomly and expect it to work.  You need to disable them via cmake, so that the code no longer has references to the specific libraries.

For instance, in the root `CMakeLists.txt` file of your game project, you can disable extensions and libraries like this:

```
set(AX_ENABLE_EXT_SPINE OFF CACHE BOOL "Build extension spine" FORCE)
set(AX_ENABLE_EXT_DRAGONBONES OFF CACHE BOOL "Build extension DragonBones" FORCE)
set(AX_ENABLE_EXT_COCOSTUDIO OFF CACHE BOOL "Build extension cocostudio" FORCE)
set(AX_ENABLE_EXT_FAIRYGUI OFF CACHE BOOL "Build extension FairyGUI" FORCE)
set(AX_ENABLE_EXT_ASSETMANAGER OFF CACHE BOOL "Build extension asset-manager" FORCE)
set(AX_ENABLE_EXT_PARTICLE3D OFF CACHE BOOL "Build extension Particle3D" FORCE)
set(AX_ENABLE_EXT_LUA OFF CACHE BOOL "Build lua libraries" FORCE)
```

Search for "AX_ENABLE_EXT" in the code for other extensions you can disable (although some are off by default).

To see where in the `CMakeLists.txt` you should add them, refer to this:
![image](https://user-images.githubusercontent.com/8603230/221441243-ba6c3b52-0061-4347-b83c-eea9c1c05251.png)

You just need to disable what you don't need for the operation of your application.

Once you change `CMakeLists.txt`, make sure you re-run the cmake command to recreate the project (for Visual Studio), and then re-build your project.

**Reply by 10100010010001001010** [reply]
Thanks, how should I connect them back if necessary?

**Reply by rh101** [reply]
I'm not sure what you mean by "connect them back".

If you mean how to re-enable those libraries, then either delete the line, or just set it to "ON" rather than "OFF".

For example, to not use the Spine library, this is OFF:
`set(AX_ENABLE_EXT_SPINE OFF CACHE BOOL "Build extension spine" FORCE)`

This is ON:
`set(AX_ENABLE_EXT_SPINE ON CACHE BOOL "Build extension spine" FORCE)`

**Reply by 10100010010001001010** [reply]
i mean if i build a .sln with the first option, then i will have a project without these extensions. How should I add these extensions to an already existing project? For example, in the future I will need Spine, but now i have clear axmol project without extensions

**Reply by rh101** [reply]
> i mean if i build a .sln with the first option, then i will have a project without these extensions. How should I add these extensions to an already existing project? For example, in the future I will need Spine, but now i have clear axmol project without extensions

You should not be editing your solution (`.sln`) or VS project files manually, and anything you need to enable or disable should be done in the `CMakeLists.txt`, and then you re-generate the Visual Studio project with cmake.  

Once again, *everything* you want to change about your project must be done via cmake (by editing your `CMakeLists.txt` and re-running cmake to create the build files), otherwise you're going to end up with a lot of problems.

**Reply by 10100010010001001010** [reply]
I understood this, but wouldn't the updated lists create a new default .sln? Or is it possible to make changes to the current project in this way? Sorry if I'm asking stupid questions, but I've never worked with cmake before

**Reply by rh101** [reply]
> I understood this, but wouldn't the updated lists create a new default .sln?

Exactly, and this is what you **want** to happen.

>Or is it possible to make changes to the current project in this way? 

Changes must **only** be made via `CMakeLists.txt`.  If you make changes to your VS solution or project files, they will be overwritten the next time you run cmake.  Any option or setting you want to change in your VS project should be done via cmake.

By using cmake, you no longer need to maintain different build and project files for different platforms, since cmake is now responsible for generating those files for you.

**Comment by halx99** [comment]
Delete `CMakeCache.txt` and regenerate .sln with cmake again, recommaded use cmake options: https://github.com/axmolengine/axmol/blob/dev/CMakeOptions.md

**Comment by 10100010010001001010** [comment]
thanks for the help everyone
